Loot/JSON serialisation improvements

Juuxel opened the pull request:
- Closes #2552.
- `class_5330$class_5331` (currently called `TypeHandler`) is really just a builder for JSON (de)serializers that are backed by a registry, so I renamed it to `SerializerBuilder`.
  - Also mapped some methods in that class
- Added `createGsonSerializer` mappings where they were missing
- `LootNbtProvider.getNbtTag` -> `getNbt` (self-explanatory, I think)
- Closes #2551

Juuxel committed:
Loot/JSON serialisation improvements

liach commented on mappings/net/minecraft/util/JsonSerializing.mapping:
```diff
@@ -1,29 +1,35 @@
+			ARG 2 serializer
+		METHOD method_33409 defaultType (Lnet/minecraft/class_5336;)Lnet/minecraft/class_5330$class_5331;
+			ARG 1 defaultType
+	CLASS class_5332 type
```
```suggestion
	CLASS class_5332 Type
```

Juuxel replied:
Yeah no this should still have the old name, not sure what I did here...

Juuxel commented on mappings/net/minecraft/util/JsonSerializing.mapping:
```diff
@@ -1,29 +1,35 @@
+			ARG 2 serializer
+		METHOD method_33409 defaultType (Lnet/minecraft/class_5336;)Lnet/minecraft/class_5330$class_5331;
+			ARG 1 defaultType
+	CLASS class_5332 type
```
```suggestion
	CLASS class_5332 CustomSerializer
```

Juuxel committed:
Update mappings/net/minecraft/util/JsonSerializing.mapping

Juuxel commented:
@liach I wonder if there should be some more renames in `JsonSerializing`:
- matching up the builder's `customSerializer` and the serializer's `elementSerializer` since they're the same thing
- renaming `CustomSerializer` to `SerializerImpl` since it's not really "custom"

Juuxel committed:
More JsonSerializing renames

Juuxel commented:
Okay, I renamed `CustomSerializer` to `ElementSerializer` and added a bit of jd for the corresponding builder method, and for clarity, `typeIdentification` to `typeGetter`.

Juuxel committed:
Clarify JD

github-actions[bot] committed:
Merge branch '1.17.1' into loot-json-stuff
